docs/develop-resources/api-conventions/validations.md

```diff
@@ -0,0 +1,158 @@
+
+See  for general reference guide;
+
+See  for parent validation with `status.externalRef`.   
+
+
+## Rule 4: No `anyOf`, `oneOf`, or` allOf`
+
+Short answer: Not now.
+
+Full answer: 
+
+OpenAPI supports `oneOf`, `anyOf` and `allOf`. Thus, in theory these should be the CRD level rules. The Direct Resource CRD is auto-generated by controller-gen, which only reads the kubebuilder tag rules. But the kubebuilder tags [do not (and most likely won’t)](https://github.com/kubernetes-sigs/controller-tools/issues/461) support `oneOf`, `anyOf` or `allOf`.
+
+To set this as CRD level rule, ConfigConnector shall build its own CRD transformer.  Ideally, this could be a ConfigConnector tag similar to kubebuilder, but this requires integrating with controller-gen which is a non-trivial amount of work. Other options like webhook, OpenAPI schema modifier are not self-explaining and easy enough to use, considering that Direct Resource shall be open to external contributors. Thus, ConfigConnector does not use CRD level validation for `oneOf`, `anyOf` and `allOf`.
+
+One future improvement can be having a Direct Resource tag that adds CRD level checks for   `oneOf`, `anyOf` or `allOf`.
+
+
+```
+// +ConfigConnector:validation:oneOf:serviceAttachmentRef,targetGRPCProxyRef,targetHTTPProxyRef,targetHTTPSProxyRef"
+type target struct {
+        ServiceAttachmentRef *refs.ComputeServiceAttachmentRef `json:"serviceAttachmentRef,omitempty"`
+
+        TargetGRPCProxyRef *refs.ComputeTargetGrpcProxyRef `json:"targetGRPCProxyRef,omitempty"`
+
+        TargetHTTPProxyRef *refs.ComputeTargetHTTPProxyRef `json:"targetHTTPProxyRef,omitempty"`
+
+        TargetHTTPSProxyRef *refs.ComputeTargetHTTPSProxyRef `json:"targetHTTPSProxyRef,omitempty"`
+}
+```
+
+
+
+# Controller validation
+
+Advanced Kubernetes contributors can use webhook to validate the resource CRs. 
+
+To easily ramp-up external contributors, we recommend using the controller-level validation instead. This validation mostly happens on Adapter [initialization](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/91aac5186eb0aa2c5c3def94d7a7c79f948ac9a9/pkg/controller/direct/directbase/directbase_controller.go#L226), GCP object [creation](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/91aac5186eb0aa2c5c3def94d7a7c79f948ac9a9/pkg/controller/direct/directbase/directbase_controller.go#L283) and [update](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/91aac5186eb0aa2c5c3def94d7a7c79f948ac9a9/pkg/controller/direct/directbase/directbase_controller.go#L291) steps. 
+
+
+## Rule 1: Resource Reference
+
+ConfigConnector shall validate the resource reference based on  
```

Comment:
Based on what..? Seems to be missing info.

Reply:
ah, looks like  Docs™ to Markdown cannot render the inline google doc. Fixed